### PR TITLE
Allow emscripten to use the full ubsan runtime

### PR DESCRIPTION
**Pull request by quantum5**
Changes:

* Made `sanitizer_common` and `ubsan` runtimes work:
  * Add `SANITIZER_EMSCRIPTEN` platform
  * Add `sanitizer_emscripten.cc` for Emscripten-specific implementations of
    helper and abstraction functions
  * Stub out functions that do not work on Emscripten. Using these will be
    abort the program. These will be implemented later.
* Add `emscripten_get_module_name` to get the WASM module name
* Changed `embuilder` to build `ubsan` and `sanitizer_common` runtimes
* Add `ubsan`-related unit tests
* Remove old JS-based `ubsan` hack and its unit tests

Known limitations:

* Stack traces do not work.

### CODE REVIEW

**Comment by aheejin** [comment]
- Which version or the llvm is this PR based on? Could you add a readme file that says this, like other libraries? And if possible, how about getting the latest release (at the moment it is 8.0.0), rather than the tip of the trunk?

- Could you point out which part of the library you had to modify to make this work? @kripken, do we have any way to say which part of the third party code was modified locally? Do we comment?

**Comment by quantum5** [comment]
I will put in the LLVM version later. As for which parts of the sanitizers were modified, see [here](https://github.com/quantum5/emscripten/compare/1e3e6de0f6bb51a7ab77846bd55f4519c7ddde75...quantum5:ubsan-full).

**Comment by aheejin** [comment]
How about getting the latest release (at the moment it is 8.0.0), rather than the tip of the trunk?

**Comment by quantum5** [comment]
The link will continue to work with new commits pushed, since it references the branch of this pull request. Unless we rebase the commits onto a different version of LLVM, of course.

**Comment by aheejin** [comment]
- Can I ask one more time, how about getting the latest release (at the moment it is 8.0.0), rather than the tip of the trunk? If you don't like the idea, could you please explain why?
- I still think splitting PR is better. Smaller PRs are generally easier to review, and especially big PR like this, having to see your custom-provided link to review a PR does not seem to be a good idea.

**Comment by quantum5** [comment]
Okay, I'll make a PR that pulls in the latest LLVM release and base this on that.

**Comment by quantum5** [comment]
This won't work until https://reviews.llvm.org/D62210 has been merged into LLVM.

**Comment by quantum5** [comment]
This should now be fully working.

File: tools/system_libs.py [hunk 1]
```diff
@@ -571,6 +571,23 @@ def create_wasm_ubsan_minimal_rt(libname):
...
... 'sanitizer_platform_limits_posix.cc', 'sanitizer_platform_limits_solaris.cc', 'sanitizer_posix.cc', 'sanitizer_posix_libcdep.cc', 'sanitizer_printf.cc', 'sanitizer_procmaps_bsd.cc', 'sanitizer_procmaps_common.cc', 'sanitizer_procmaps_linux.cc', 'sanitizer_procmaps_mac.cc', 'sanitizer_procmaps_solaris.cc', 'sanitizer_rtems.cc', 'sanitizer_solaris.cc', 'sanitizer_stackdepot.cc', 'sanitizer_stacktrace.cc', 'sanitizer_stacktrace_libcdep.cc', 'sanitizer_stacktrace_printer.cc', 'sanitizer_stacktrace_sparc.cc', 'sanitizer_stoptheworld_linux_libcdep.cc', 'sanitizer_stoptheworld_mac.cc', 'sanitizer_suppressions.cc', 'sanitizer_symbolizer.cc', 'sanitizer_symbolizer_libbacktrace.cc', 'sanitizer_symbolizer_libcdep.cc', 'sanitizer_symbolizer_mac.cc', 'sanitizer_symbolizer_markup.cc', 'sanitizer_symbolizer_posix_libcdep.cc', 'sanitizer_symbolizer_report.cc', 'sanitizer_symbolizer_win.cc', 'sanitizer_termination.cc', 'sanitizer_thread_registry.cc', 'sanitizer_tls_get_addr.cc', 'sanitizer_type_traits.cc', 'sanitizer_unwind_linux_libcdep.cc', 'sanitizer_unwind_win.cc', 'sanitizer_win.cc', 'sanitizer_win_dll_thunk.cc', 'sanitizer_win_dynamic_runtime_thunk.cc', 'sanitizer_win_weak_interception.cc'],
```
Comment (tlively) [inline tools/system_libs.py]: This looks rather brittle. Is there a more general way to list these files?

File: tools/system_libs.py [hunk 2]
```diff
@@ -783,6 +800,13 @@ def do_create():
 
   if shared.Settings.UBSAN_RUNTIME == 1:
     libs_to_link.append((shared.Cache.get('libubsan_minimal_rt_wasm.a', lambda: create_wasm_ubsan_minimal_rt('libubsan_minimal_rt_wasm.a')), False))
+  elif shared.Settings.UBSAN_RUNTIME == 2:
+    libs_to_link.append((shared.Cache.get('libsanitizer_common_rt_wasm.a', lambda: create_wasm_common_san_rt('libsanitizer_common_rt_wasm.a')), False))
+    libs_to_link.append((shared.Cache.get('libubsan_rt_wasm.a', lambda: create_wasm_ubsan_rt('libubsan_rt_wasm.a')), False))
+    # TODO: handle this dependency better
```
Comment (tlively) [inline tools/system_libs.py]: Can you elaborate on the problem in this TODO?

File: tests/test_other.py [hunk 3]
```diff
@@ -7971,7 +7932,7 @@ def test_binaryen_metadce_cxx_fastcomp(self, *args):
     # don't compare the # of functions in a main module, which changes a lot
     # TODO(sbc): Investivate why the number of exports is order of magnitude
     # larger for wasm backend.
-    'main_module_1': (['-O3', '-s', 'MAIN_MODULE=1'], 1581, [], [], 517336, 172, 1484, None), # noqa
+    'main_module_1': (['-O3', '-s', 'MAIN_MODULE=1'], 1582, [], [], 517336, 172, 1484, None), # noqa
```
Comment (tlively) [inline tests/test_other.py]: Why did these numbers change?

File: system/lib/compiler-rt/lib/ubsan/ubsan_signals_standalone.cc [hunk 4]
```diff
@@ -28,13 +23,18 @@
...
 // it is part of an APK that does not use wrap.sh method.
-#if SANITIZER_FUCHSIA || SANITIZER_ANDROID
+#if SANITIZER_FUCHSIA || SANITIZER_ANDROID || SANITIZER_EMSCRIPTEN
 
 namespace __ubsan {
 void InitializeDeadlySignals() {}
 }
 
-#else
+#elif CAN_SANITIZE_UB
```
Comment (tlively) [inline system/lib/compiler-rt/lib/ubsan/ubsan_signals_standalone.cc]: Why did this chunk of code  have to move?

File: system/lib/compiler-rt/lib/sanitizer_common/sanitizer_symbolizer_report.cc [hunk 5]
```diff
@@ -40,7 +40,12 @@ void ReportErrorSummary(const char *error_type, const AddressInfo &info,
 }
 #endif
 
-#if !SANITIZER_FUCHSIA
+#if SANITIZER_EMSCRIPTEN
+
+// Most JavaScript consoles do not handle colors.
+static INLINE bool ReportSupportsColors() { return false; }
```
Comment (tlively) [inline system/lib/compiler-rt/lib/sanitizer_common/sanitizer_symbolizer_report.cc]: :+1:

File: system/lib/compiler-rt/lib/sanitizer_common/sanitizer_posix.cc [hunk 6]
```diff
@@ -278,6 +289,11 @@ void ReportFile::Write(const char *buffer, uptr length) {
   internal_write(fd, buffer, length);
 }
 
+#if SANITIZER_EMSCRIPTEN
+bool GetCodeRangeForFile(const char */*module*/, uptr */*start*/, uptr */*end*/) {
+  return false;
```
Comment (tlively) [inline system/lib/compiler-rt/lib/sanitizer_common/sanitizer_posix.cc]: It would be good to add a short comment here.

File: emcc.py [hunk 7]
```diff
@@ -115,6 +115,43 @@
 # Target options
 final = None
 
+UBSAN_SANITIZES = {
```
Comment (tlively) [inline emcc.py]:
```suggestion
UBSAN_SANITIZERS = {
```

File: emcc.py [hunk 8]
```diff
@@ -1423,8 +1460,19 @@ def check(input_file):
...
-        shared.Settings.UBSAN_RUNTIME = 1
+      sanitize = set()
+
+      for arg in newargs:
+        if arg.startswith('-fsanitize='):
+          sanitize.update(arg.split('=', 1)[1].split(','))
+        elif arg.startswith('-fno-sanitize='):
+          sanitize.difference_update(arg.split('=', 1)[1].split(','))
+
+      if sanitize & UBSAN_SANITIZES:
```
Comment (tlively) [inline emcc.py]:
```suggestion
      if sanitize & UBSAN_SANITIZERS:
```

File: embuilder.py [hunk 9]
```diff
@@ -313,6 +314,10 @@ def is_flag(arg):
         }
 
       ''', [libname('libpthreads_stub')])
+    elif what == 'ubsan-rt-wasm':
+      if not shared.Settings.WASM_BACKEND:
+        continue
+      build(C_BARE, ['libsanitizer_common_rt_wasm.a', 'libubsan_rt_wasm.a'], ['-fsanitize=undefined', '-s', 'WASM=1'])
```
Comment (tlively) [inline embuilder.py]: It looks like you're building the sanitizer code with ubsan enabled. Is that necessary? Shouldn't the sanitizer code be trusted to run correctly?

File: src/library_syscall.js [hunk 10]
```diff
@@ -224,6 +224,9 @@ var SyscallsLibrary = {
     exit(status);
     return 0;
   },
+  __syscall2: function(which, varargs) { // fork
+    return -{{{ cDefine('ENOSYS') }}};
+  },
```
Comment (tlively) [inline src/library_syscall.js]: Why is this necessary now but not before? It seems like it would be a better idea to remove or #ifdef out any code that wants to call fork.

File: src/library_syscall.js [hunk 11]
```diff
@@ -910,6 +913,10 @@ var SyscallsLibrary = {
   __syscall153: function(which, varargs) { // munlockall
     return 0;
   },
+  __syscall158: function(which, varargs) { // sched_yield
+    // always succeeds, not guaranteed to do anything anyways
+    return 0;
```
Comment (tlively) [inline src/library_syscall.js]: This may be legal, but will it cause surprising behavior? Would it make more sense to #ifdef out any code that expects to use it?

File: src/library_syscall.js [hunk 12]
```diff
@@ -1175,6 +1183,14 @@ var SyscallsLibrary = {
     }
 #endif // SYSCALLS_REQUIRE_FILESYSTEM
   },
+  __syscall240: function(which, varargs) { // futex
+    return -{{{ cDefine('ENOSYS')}}};
+  },
```
Comment (tlively) [inline src/library_syscall.js]: Same question for futex; would it make more sense to #ifdef out code that uses it?

File: system/lib/compiler-rt/lib/sanitizer_common/sanitizer_posix.cc [hunk 13]
```diff
@@ -279,6 +290,10 @@ void ReportFile::Write(const char *buffer, uptr length) {
 }
 
 bool GetCodeRangeForFile(const char *module, uptr *start, uptr *end) {
+#if SANITIZER_EMSCRIPTEN
+  // Code is not mapped in memory in Emscripten, so this operation is meaningless
+  // and thus always fails.
```
Comment (tlively) [inline system/lib/compiler-rt/lib/sanitizer_common/sanitizer_posix.cc]: Would it make more sense to always succeed with an empty range?

**Comment by aheejin** [comment]
Please don't force-push commits.
From an [article](https://help.github.com/en/articles/about-pull-requests) from the github blog:
> When pushing commits to a pull request, don't force push. Force pushing can corrupt your pull request.

**Comment by quantum5** [comment]
The last force push was necessary because git failed to rebase a merge commit properly, resulting in all the changes on the incoming branch showing up in the diff.

File: system/lib/compiler-rt/ubsan-minimal.symbols [hunk 14]
```diff
@@ -0,0 +1,79 @@
...
+-------- T __ubsan_handle_sub_overflow_minimal
+-------- T __ubsan_handle_sub_overflow_minimal_abort
+-------- T __ubsan_handle_type_mismatch_minimal
+-------- T __ubsan_handle_type_mismatch_minimal_abort
+-------- T __ubsan_handle_vla_bound_not_positive_minimal
+-------- T __ubsan_handle_vla_bound_not_positive_minimal_abort
+         U abort
+         U emscripten_return_address
+         U strlen
+         U write
```
Comment (tlively) [inline system/lib/compiler-rt/ubsan-minimal.symbols]: What is this file for?

**Comment by aheejin** [comment]
> The last force push was necessary because git failed to rebase a merge commit properly, resulting in all the changes on the incoming branch showing up in the diff.

Does that happen when you do `git merge` and not `git rebase`?

File: system/lib/compiler-rt/lib/ubsan/readme.txt [hunk 15]
```diff
@@ -1 +1,7 @@
 These files are from compiler-rt 8.0.0.
+
+===========================================================================
+
+Changes were made to make this work with Emscripten:
+
+https://github.com/quantum5/emscripten/compare/ubsan-full-upstream...ubsan-full
```
Comment (aheejin) [inline system/lib/compiler-rt/lib/ubsan/readme.txt]: The same; let's not put personal repo links here.

File: system/lib/compiler-rt/ubsan-minimal.symbols (hunk 14 above)
Comment (quantum5) [reply system/lib/compiler-rt/ubsan-minimal.symbols]: I converted ubsan to be a system library instead of just `libraries_to_link`, and that requires symbol files. I did the same thing for the minimal runtime to be consistent, and this is the result.

File: system/lib/compiler-rt/lib/sanitizer_common/sanitizer_posix.cc (hunk 13 above)
Comment (quantum5) [reply system/lib/compiler-rt/lib/sanitizer_common/sanitizer_posix.cc]: I am more inclined to make this an error. We don't know what this is used for yet, so I'd rather not make it succeed.

File: ChangeLog.md [hunk 16]
```diff
@@ -17,6 +17,9 @@ Current Trunk
 -------------
  - Add `emscripten_return_address` which implements the functionality of
    gcc/clang's `__builtin_return_address`. (#8617)
+ - Add support for [undefined behavior sanitizer](https://clang.llvm.org/docs/UndefinedBehaviorSanitizer.html).
+   This allows `emcc -fsanitize=undefined` to work. (#8651)
+   The minimal runtime (`-fsanitize-minimal-runtime`) also works. (#8617)
```
Comment (aheejin) [inline ChangeLog.md]: Without sub-bullets, these two lines will be merged into the previous line. You can do like
```
 - Add support for [undefined behavior sanitizer]
   - This allows `emcc -fsanitize=undefined` to work. (#8651)
   - The minimal runtime (`-fsanitize-minimal-runtime`) also works. (#8617)
```

File: emcc.py [hunk 17]
```diff
@@ -1423,8 +1460,19 @@ def check(input_file):
...
-        shared.Settings.UBSAN_RUNTIME = 1
+      sanitize = set()
+
+      for arg in newargs:
+        if arg.startswith('-fsanitize='):
+          sanitize.update(arg.split('=', 1)[1].split(','))
+        elif arg.startswith('-fno-sanitize='):
+          sanitize.difference_update(arg.split('=', 1)[1].split(','))
+
+      if sanitize & UBSAN_SANITIZERS:
```
Comment (kripken) [inline emcc.py]: why does `&` work here?

regardless, I am worried about having that `UBSAN_SANITIZERS` list, and all the duplication of parsing logic using that. Can we link in ubsan based on undefined symbols, like we do other  things?

Comment (quantum5) [reply emcc.py]: `&` is overloaded as set intersection. As for undefined symbols, I am going to try making that work.

File: tools/system_libs.py [hunk 18]
```diff
@@ -744,6 +779,9 @@ def do_create():
     for d in lib.deps:
       add_library(system_libs_map[d])
 
+    for d in lib.js_deps:
+      shared.Settings.EXPORTED_FUNCTIONS.append('_%s' % d)
```
Comment (sbc100) [inline tools/system_libs.py]: Can you explain this a litte? It looks like it just used for memalign?  Is memalign implemented in JS?   Why to do need js_deps now but we've managed to get away without them previously?

File: tools/system_libs.py [hunk 19]
```diff
@@ -781,9 +819,6 @@ def do_create():
     libs_to_link.append((shared.Cache.get('libcompiler_rt_wasm.a', lambda: create_wasm_compiler_rt('libcompiler_rt_wasm.a')), False))
     libs_to_link.append((shared.Cache.get('libc_rt_wasm.a', lambda: create_wasm_libc_rt('libc_rt_wasm.a')), False))
 
-  if shared.Settings.UBSAN_RUNTIME == 1:
-    libs_to_link.append((shared.Cache.get('libubsan_minimal_rt_wasm.a', lambda: create_wasm_ubsan_minimal_rt('libubsan_minimal_rt_wasm.a')), False))
-
```
Comment (sbc100) [inline tools/system_libs.py]: How does ubsan work on regular systems?  Presumably you need to add a linker flag otherwise you'll get undefined symbols?   I guess we can use our dependency system to do one better but I think requiring the linker flag is also ok.

File: system/lib/compiler-rt/sanitizer-common.symbols [hunk 20]
```diff
@@ -0,0 +1,2058 @@
...
+-------- T _ZN11__sanitizer16internal_strncatEPcPKcm
+-------- T _ZN11__sanitizer16internal_strncmpEPKcS1_m
+-------- T _ZN11__sanitizer16internal_strncpyEPcPKcm
+-------- T _ZN11__sanitizer16internal_strnlenEPKcm
+-------- T _ZN11__sanitizer16internal_strrchrEPKci
+-------- T _ZN11__sanitizer18internal_strchrnulEPKci
+-------- T _ZN11__sanitizer23internal_simple_strtollEPKcPS1_i
+
+sanitizer_allocator.cc.o:
+-------- d .L.str
```
Comment (sbc100) [inline system/lib/compiler-rt/sanitizer-common.symbols]: I don't see these local symbols on the other `.symbols` file.  Is there something we are doing differently here?

File: tools/system_libs.py (hunk 19 above)
Comment (quantum5) [reply tools/system_libs.py]: `clang` automatically adds the ubsan library to the linker command line.

File: tools/system_libs.py (hunk 18 above)
Comment (quantum5) [reply tools/system_libs.py]: This is `__syscall192` (a JS function) depending on `memalign` (a C function), but we only add back dependencies to user code, not libraries.

Here, the sanitizer common library depends on `__syscall192`, but `memalign` is not pulled in because the user code doesn't use `__syscall192`.

I am not sure if this is the best approach, but I am doing this for now so that it at least works.

File: tools/system_libs.py (hunk 19 above)
Comment (sbc100) [reply tools/system_libs.py]: But only if you specify the `-fsanitize..` flag no?

File: system/lib/compiler-rt/sanitizer-common.symbols (hunk 20 above)
Comment (quantum5) [reply system/lib/compiler-rt/sanitizer-common.symbols]: I am not sure how these files are generated. I just did `llvm-nm` and replaced the addresses with `-` with `perl`.

File: tools/system_libs.py (hunk 19 above)
Comment (quantum5) [reply tools/system_libs.py]: Yea. The code you commented on was the old code that parses the command line for `-fsanitize`, but @kripken wanted to use the dependency system, and so it's now doing that.

**Comment by tlively** [comment]
@sbc100 do you have any other concerns?

File: ChangeLog.md [hunk 21]
```diff
@@ -23,6 +23,10 @@ v1.38.33: 05/23/2019
 
  - Add `emscripten_return_address` which implements the functionality of
    gcc/clang's `__builtin_return_address`. (#8617)
+ - Add support for [undefined behavior sanitizer](https://clang.llvm.org/docs/UndefinedBehaviorSanitizer.html).
```
Comment (sbc100) [inline ChangeLog.md]: Should this code in the current trunk section above?

File: emcc.py [hunk 22]
```diff
@@ -1423,9 +1423,6 @@ def check(input_file):
       if options.separate_asm:
         exit_with_error('cannot --separate-asm when emitting wasm, since not emitting asm.js')
 
-      if '-fsanitize-minimal-runtime' in newargs:
-        shared.Settings.UBSAN_RUNTIME = 1
```
Comment (sbc100) [inline emcc.py]: Can we completely remove this setting then?

File: system/lib/compiler-rt/lib/sanitizer_common/sanitizer_emscripten.cc [hunk 23]
```diff
@@ -0,0 +1,62 @@
...
+
+char *fake_argv[] = {0};
+char *fake_envp[] = {0};
+
+char **GetArgv() {
+  return fake_argv;
+}
+
+char **GetEnviron() {
+  return fake_envp;
```
Comment (sbc100) [inline system/lib/compiler-rt/lib/sanitizer_common/sanitizer_emscripten.cc]: Why fake this?  Can't you use emscripten_get_env()?

File: system/lib/compiler-rt/ubsan-minimal.symbols (hunk 14 above)
Comment (sbc100) [reply system/lib/compiler-rt/ubsan-minimal.symbols]: Again, I think if we revert to the old behavior of linking if and only if the user specific `-fsantize` we can avoid these extra symbols files.   Maybe wait for @kripken to respond because I know he explicitly advised you to go this route.

File: emcc.py (hunk 22 above)
Comment (quantum5) [reply emcc.py]: Yes, that was the intention. I thought I did it already. It will be done in the next update.

File: system/lib/compiler-rt/lib/sanitizer_common/sanitizer_emscripten.cc (hunk 23 above)
Comment (quantum5) [reply system/lib/compiler-rt/lib/sanitizer_common/sanitizer_emscripten.cc]: For some strange reason, the sanitizers want a function to get the entire environment block, as in a sequence of null-terminated strings going `NAME=VALUE`. It doesn't appear to be used anywhere that makes a difference yet, so I am stubbing it out.

File: ChangeLog.md (hunk 21 above)
Comment (quantum5) [reply ChangeLog.md]: Thanks for catching this. Git merge messed this up.

**Comment by kripken** [comment]
I remove my objection to duplicating the sanitizer flags, if you and @sbc100 both think that's better. So can avoid using the undefined symbols.

File: emcc.py [hunk 24]
```diff
@@ -115,6 +115,43 @@
...
+  'vla-bound',
+  'vptr',
+  'undefined',
+  'undefined-trap',
+  'implicit-integer-truncation',
+  'implicit-integer-arithmetic-value-change',
+  'implicit-conversion',
+  'integer',
+  'nullability',
+}
```
Comment (sbc100) [inline emcc.py]: Oh nice, set literals!  We should probably use them more in emscripten.

File: system/lib/compiler-rt/sanitizer-common.symbols [hunk 25]
```diff
@@ -1,1392 +0,0 @@
-
```
Comment (sbc100) [inline system/lib/compiler-rt/sanitizer-common.symbols]: Can you delete these .symbols files now?

Comment (quantum5) [reply system/lib/compiler-rt/sanitizer-common.symbols]: They have been deleted.

File: src/library.js [hunk 26]
```diff
@@ -4677,6 +4679,18 @@ LibraryManager.library = {
     }
   },
 
+  emscripten_get_module_name: function(buf, length) {
+    return stringToUTF8(wasmBinaryFile, buf, length);
```
Comment (kripken) [inline src/library.js]: Sorry for the late review!

`wasmBinaryFile` will not exist in wasm2js mode, or if the wasm was embedded in the JS. But maybe it's not worth trying to support more cases than the common one.

Comment (quantum5) [reply src/library.js]: It appears to exist in wasm2js mode, which is why I did this:

```console
$ emcc -fsanitize=null null-read.c -s WASM=0
emcc:WARNING: emcc: JS support in the upstream LLVM+wasm2js path is very experimental currently (best to use fastcomp for asm.js for now)
emscripten:WARNING: Wasm source map won't be usable in a browser without --source-map-base
$ grep wasmBinaryFile a.out.js 
var wasmBinaryFile = 'a.out.wasm';
if (!isDataURI(wasmBinaryFile)) {
  wasmBinaryFile = locateFile(wasmBinaryFile);
      return Module['readBinary'](wasmBinaryFile);
    return fetch(wasmBinaryFile, { credentials: 'same-origin' }).then(function(response) {
        throw "failed to load wasm binary file at '" + wasmBinaryFile + "'";
        !isDataURI(wasmBinaryFile) &&
      return WebAssembly.instantiateStreaming(fetch(wasmBinaryFile, { credentials: 'same-origin' }), info)
      return stringToUTF8(wasmBinaryFile, buf, length);
$ node a.out.js 
null-read.c:3:9: runtime error: load of null pointer of type 'int'
```

Comment (kripken) [reply src/library.js]: I don't understand? The file doesn't exist, but the variable does, yes. Is just the variable name enough?

Comment (quantum5) [reply src/library.js]: Yes, it just wants a name.
